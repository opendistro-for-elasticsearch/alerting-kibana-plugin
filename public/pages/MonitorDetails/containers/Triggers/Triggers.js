/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import uuidv4 from 'uuid/v4';
import { EuiButton, EuiInMemoryTable } from '@elastic/eui';

import ContentPanel from '../../../../components/ContentPanel';

const MAX_TRIGGERS = 10;

export default class Triggers extends Component {
  constructor(props) {
    super(props);

    this.state = {
      field: 'name',
      tableKey: uuidv4(),
      direction: 'asc',
      selectedItems: [],
    };

    this.onDelete = this.onDelete.bind(this);
    this.onEdit = this.onEdit.bind(this);
    this.onSelectionChange = this.onSelectionChange.bind(this);
    this.onTableChange = this.onTableChange.bind(this);
  }

  componentWillReceiveProps(nextProps, nextContext) {
    if (this.props.monitor !== nextProps.monitor) {
      // In the React version Kibana uses there is a bug regarding getDerivedStateFromProps
      // which EuiInMemoryTable uses which causes items to not be updated correctly.
      // Whenever the monitor is updated we'll generate a new key for the table
      // which will cause the table component to remount
      this.setState({ tableKey: uuidv4() });
    }
  }

  onDelete() {
    const { selectedItems } = this.state;
    const { updateMonitor, monitor } = this.props;
    const isAggregationMonitor = monitor.monitor_type === 'aggregation_monitor';
    const triggersToDelete = selectedItems.reduce(
      (map, item) => ({
        ...map,
        [item.name]: true,
      }),
      {}
    );
    const shouldKeepTrigger = (trigger) => !triggersToDelete[trigger.name];
    const updatedTriggers = this.getUnwrappedTriggers(monitor).filter(shouldKeepTrigger);
    updateMonitor({ triggers: updatedTriggers });
  }

  onEdit() {
    // TODO: Rewrapping the Trigger onEdit to avoid complicating triggerToFormik for now
    const { monitor } = this.props;
    const selectedTrigger = this.state.selectedItems[0];
    const triggerType =
      monitor.monitor_type === 'aggregation_monitor'
        ? 'aggregation_trigger'
        : 'traditional_trigger';
    const wrappedTrigger = {
      [triggerType]: selectedTrigger,
    };
    this.props.onEditTrigger(wrappedTrigger);
  }

  onSelectionChange(selectedItems) {
    this.setState({ selectedItems });
  }

  onTableChange({ sort: { field, direction } = {} }) {
    this.setState({ field, direction });
  }

  // TODO: For now, unwrapping all the Triggers since it's conflicting with the table
  //   retrieving the 'id' and causing it to behave strangely
  getUnwrappedTriggers(monitor) {
    const isAggregationMonitor = monitor.monitor_type === 'aggregation_monitor';
    return isAggregationMonitor
      ? monitor.triggers.map((trigger) => {
          return trigger.aggregation_trigger;
        })
      : monitor.triggers.map((trigger) => {
          return trigger.traditional_trigger;
        });
  }

  render() {
    const { direction, field, selectedItems, tableKey } = this.state;
    const { monitor, onCreateTrigger } = this.props;

    const columns = [
      {
        field: 'name',
        name: 'Name',
        sortable: true,
        truncateText: true,
      },
      {
        field: 'actions',
        name: 'Number of actions',
        sortable: true,
        truncateText: false,
        render: (actions) => actions.length,
      },
      {
        field: 'severity',
        name: 'Severity',
        sortable: true,
        truncateText: false,
      },
    ];

    const sorting = { sort: { field, direction } };

    const selection = { onSelectionChange: this.onSelectionChange };

    return (
      <ContentPanel
        title="Triggers"
        titleSize="s"
        bodyStyles={{ padding: 'initial' }}
        actions={[
          <EuiButton onClick={this.onEdit} disabled={selectedItems.length !== 1}>
            Edit
          </EuiButton>,
          <EuiButton onClick={this.onDelete} disabled={!selectedItems.length}>
            Delete
          </EuiButton>,
          <EuiButton
            onClick={onCreateTrigger}
            disabled={monitor.triggers.length >= MAX_TRIGGERS}
            fill
          >
            Create
          </EuiButton>,
        ]}
      >
        <EuiInMemoryTable
          items={this.getUnwrappedTriggers(monitor)}
          itemId="id"
          key={tableKey}
          columns={columns}
          sorting={sorting}
          isSelectable={true}
          selection={selection}
          onTableChange={this.onTableChange}
        />
      </ContentPanel>
    );
  }
}

Triggers.propTypes = {
  monitor: PropTypes.object.isRequired,
  updateMonitor: PropTypes.func.isRequired,
  onEditTrigger: PropTypes.func.isRequired,
  onCreateTrigger: PropTypes.func.isRequired,
};
