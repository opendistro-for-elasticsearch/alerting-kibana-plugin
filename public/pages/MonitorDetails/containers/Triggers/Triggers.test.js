/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';

import Triggers from './Triggers';

const props = {
  monitor: {
    triggers: [{ name: 'Random Trigger', severity: 1, actions: [{ name: 'Random Action' }] }],
  },
  updateMonitor: jest.fn(),
  onEditTrigger: jest.fn(),
  onCreateTrigger: jest.fn(),
};

jest.mock('uuid/v4', () => {
  let value = 0;
  return () => value++;
});

function getShallowWrapper(customProps = {}) {
  return shallow(<Triggers {...props} {...customProps} />);
}

describe('Triggers', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  test('renders', () => {
    const wrapper = getShallowWrapper();

    expect(wrapper).toMatchSnapshot();
  });

  test('sets new tableKey on new monitor', () => {
    const wrapper = getShallowWrapper();
    const tableKey = wrapper.instance().state.tableKey;
    wrapper.setProps({ random: 5 });
    const sameTableKey = wrapper.instance().state.tableKey;
    expect(tableKey).toBe(sameTableKey);
    wrapper.setProps({ monitor: { ...props.monitor, name: 'New Random Monitor' } });
    const diffTableKey = wrapper.instance().state.tableKey;
    expect(tableKey).not.toBe(diffTableKey);
  });

  test('onEdit calls onEditTrigger', () => {
    const onEdit = jest.spyOn(Triggers.prototype, 'onEdit');
    const wrapper = getShallowWrapper();
    wrapper.instance().onEdit();
    expect(onEdit).toHaveBeenCalled();
    expect(props.onEditTrigger).toHaveBeenCalled();
  });

  test('onDelete calls updateMonitor with triggers to keep', () => {
    const onDelete = jest.spyOn(Triggers.prototype, 'onDelete');
    const monitor = {
      triggers: [
        { name: 'one', severity: 1, actions: [{ name: 'one action' }] },
        { name: 'two', severity: 2, actions: [{ name: 'two action' }] },
      ],
    };
    const wrapper = getShallowWrapper({ monitor });
    wrapper.setState({ selectedItems: [monitor.triggers[0]] });
    wrapper.instance().onDelete();
    expect(onDelete).toHaveBeenCalled();
    expect(props.updateMonitor).toHaveBeenCalled();
    expect(props.updateMonitor).toHaveBeenCalledWith({ triggers: [monitor.triggers[1]] });
  });
});
