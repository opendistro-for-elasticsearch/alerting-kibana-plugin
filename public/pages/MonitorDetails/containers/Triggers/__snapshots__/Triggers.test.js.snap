// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Triggers renders 1`] = `
<ContentPanel
  actions={
    Array [
      <EuiButton
        color="primary"
        disabled={true}
        fill={false}
        iconSide="left"
        onClick={[Function]}
        size="m"
        type="button"
      >
        Edit
      </EuiButton>,
      <EuiButton
        color="primary"
        disabled={true}
        fill={false}
        iconSide="left"
        onClick={[Function]}
        size="m"
        type="button"
      >
        Delete
      </EuiButton>,
      <EuiButton
        color="primary"
        disabled={false}
        fill={true}
        iconSide="left"
        onClick={[MockFunction]}
        size="m"
        type="button"
      >
        Create
      </EuiButton>,
    ]
  }
  bodyStyles={
    Object {
      "padding": "initial",
    }
  }
  title="Triggers"
  titleSize="s"
>
  <EuiInMemoryTable
    columns={
      Array [
        Object {
          "field": "name",
          "name": "Name",
          "sortable": true,
          "truncateText": true,
        },
        Object {
          "field": "actions",
          "name": "Number of actions",
          "render": [Function],
          "sortable": true,
          "truncateText": false,
        },
        Object {
          "field": "severity",
          "name": "Severity",
          "sortable": true,
          "truncateText": false,
        },
      ]
    }
    executeQueryOptions={Object {}}
    isSelectable={true}
    itemId="id"
    items={
      Array [
        Object {
          "actions": Array [
            Object {
              "name": "Random Action",
            },
          ],
          "name": "Random Trigger",
          "severity": 1,
        },
      ]
    }
    key="0"
    onTableChange={[Function]}
    pagination={false}
    responsive={true}
    selection={
      Object {
        "onSelectionChange": [Function],
      }
    }
    sorting={
      Object {
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
      }
    }
  />
</ContentPanel>
`;
